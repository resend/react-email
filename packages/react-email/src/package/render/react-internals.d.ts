declare module 'react-dom/server.browser' {
  export * from 'react-dom/server';
}
