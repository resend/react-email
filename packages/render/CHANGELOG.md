# @react-email/render

## 1.0.5

### Patch Changes

- caa49b3: Fix extra `<` characters being kept when rendering if mso comments under certain conditions
- 337ac4e: Fix pretty option breaking button components
- b44c937: Fix unstable rendering when prettifying with the Preview component

## 1.0.5-canary.1

### Patch Changes

- dfc49b5: Fix extra `<` characters being kept when rendering if mso comments under certain conditions
- 301f54b: Fix unstable rendering when prettifying with the Preview component

## 1.0.5-canary.0

### Patch Changes

- ed2fa33: Fix pretty option breaking button components

## 1.0.4

### Patch Changes

- 1ff196a: Update prettier dependency.

## 1.0.3

### Patch Changes

- 467af4e: Use peerDependencies for react and react-dom
- b34aa90: Move react and react-dom to just dependencies for better DX
- da3e719: Use prettier's stadalone API instead of js-beautify
- fd3b9de: Remove uncessary destructuring of react-dom/server

## 1.0.3-canary.3

### Patch Changes

- 467af4e: Use peerDependencies for react and react-dom

## 1.0.3-canary.2

### Patch Changes

- 2ebf17f: Use prettier's stadalone API instead of js-beautify

## 1.0.3-canary.1

### Patch Changes

- 0f32e50: Remove uncessary destructuring of react-dom/server

## 1.0.3-canary.0

### Patch Changes

- f7833da: Move react and react-dom to just dependencies for better DX

## 1.0.2

### Patch Changes

- 4627675: Fix null characters in between chunks when using high-density characters

## 1.0.2-canary.0

### Patch Changes

- 0fab161: Fix null characters in between chunks when using high-density characters

## 1.0.1

### Patch Changes

- 7481b12: Add a wrapping Suspense to all email templates before rendering

## 1.0.0

### Major Changes

- f9483ec: Deprecated `renderAsync` and made `render` itself always async

  ## Why

  Three reasons:

  1. Better support of NextJS's latest versions
  2. Being ready for future React API deprecations
  3. Support for Suspense which allows for using async inside components

  See https://github.com/resend/react-email/discussions/1144 for more info.

  ## How to upgrade

  If you are using the old `render`, you will need to now treat the Promise
  that comes out of it, as it is now async. If you are using `renderAsync`,
  you can replace it with `render` and things should work the same.

### Patch Changes

- 3caaf53: Updated peer dependencies to allow for React 19 release candidated and React 19 itself

## 1.0.0-canary.1

### Major Changes

- 3f67038: Deprecated `renderAsync` and made `render` itself always async

  ## Why

  Three reasons:

  1. Better support of NextJS's latest versions
  2. Being ready for future React API deprecations
  3. Support for Suspense which allows for using async inside components

  See https://github.com/resend/react-email/discussions/1144 for more info.

  ## How to upgrade

  If you are using the old `render`, you will need to now treat the Promise
  that comes out of it, as it is now async. If you are using `renderAsync`,
  you can replace it with `render` and things should work the same.

## 0.0.18-canary.0

### Patch Changes

- a1c016b: Updated peer dependencies to allow for React 19 release candidated and React 19 itself
