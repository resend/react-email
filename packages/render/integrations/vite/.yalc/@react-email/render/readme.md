![React Email button cover](https://react.email/static/covers/render.png)

<div align="center"><strong>@react-email/render</strong></div>
<div align="center">Transform React components into HTML email templates.</div>
<br />
<div align="center">
<a href="https://react.email">Website</a> 
<span> · </span>
<a href="https://github.com/resend/react-email">GitHub</a> 
<span> · </span>
<a href="https://react.email/discord">Discord</a>
</div>

## Install

Install component from your command line.

#### With yarn

```sh
yarn add @react-email/render -E
```

#### With npm

```sh
npm install @react-email/render -E
```

## Getting started

Convert React components into a HTML string.

```jsx
import { MyTemplate } from "../components/MyTemplate";
import { render } from "@react-email/render";

const html = await render(<MyTemplate firstName="Jim" />);
```

## License

MIT License
