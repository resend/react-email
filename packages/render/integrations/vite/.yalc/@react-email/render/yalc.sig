6e502c45a0cc6d24195e55ec1ab3474c