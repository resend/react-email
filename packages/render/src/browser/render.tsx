import { convert } from 'html-to-text';
import { Suspense } from 'react';
import type {
  PipeableStream,
  ReactDOMServerReadableStream,
} from 'react-dom/server';
import type { Options } from '../shared/options';
import { plainTextSelectors } from '../shared/plain-text-selectors';
import { pretty } from '../shared/utils/pretty';

const decoder = new TextDecoder('utf-8');

const readStream = async (
  stream: PipeableStream | ReactDOMServerReadableStream,
) => {
  const chunks: Uint8Array[] = [];

  if ('pipeTo' in stream) {
    // means it's a readable stream
    const writableStream = new WritableStream({
      write(chunk: Uint8Array) {
        chunks.push(chunk);
      },
    });
    await stream.pipeTo(writableStream);
  } else {
    throw new Error(
      'For some reason, the Node version of `react-dom/server` has been imported instead of the browser one.',
      {
        cause: {
          stream,
        },
      },
    );
  }

  let length = 0;
  chunks.forEach((item) => {
    length += item.length;
  });
  const mergedChunks = new Uint8Array(length);
  let offset = 0;
  chunks.forEach((item) => {
    mergedChunks.set(item, offset);
    offset += item.length;
  });

  return decoder.decode(mergedChunks);
};

export const render = async (
  element: React.ReactElement,
  options?: Options,
) => {
  const suspendedElement = <Suspense>{element}</Suspense>;
  const reactDOMServer = await import('react-dom/server');

  let html!: string;
  if (Object.hasOwn(reactDOMServer, 'renderToReadableStream')) {
    html = await readStream(
      await reactDOMServer.renderToReadableStream(suspendedElement),
    );
  } else {
    await new Promise<void>((resolve, reject) => {
      const stream = reactDOMServer.renderToPipeableStream(suspendedElement, {
        async onAllReady() {
          html = await readStream(stream);
          resolve();
        },
        onError(error) {
          reject(error as Error);
        },
      });
    });
  }

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, '')}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
};
