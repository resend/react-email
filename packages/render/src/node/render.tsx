import { convert } from 'html-to-text';
import { Suspense } from 'react';
import type { Options } from '../shared/options';
import { plainTextSelectors } from '../shared/plain-text-selectors';
import { pretty } from '../shared/utils/pretty';
import { readStream } from './read-stream';

export const render = async (
  element: React.ReactElement,
  options?: Options,
) => {
  const suspendedElement = <Suspense>{element}</Suspense>;
  const reactDOMServer = await import('react-dom/server');

  let html!: string;
  if (Object.hasOwn(reactDOMServer, 'renderToReadableStream')) {
    html = await readStream(
      await reactDOMServer.renderToReadableStream(suspendedElement),
    );
  } else {
    await new Promise<void>((resolve, reject) => {
      const stream = reactDOMServer.renderToPipeableStream(suspendedElement, {
        async onAllReady() {
          html = await readStream(stream);
          resolve();
        },
        onError(error) {
          reject(error as Error);
        },
      });
    });
  }

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, '')}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
};
