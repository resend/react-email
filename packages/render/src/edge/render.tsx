import { Suspense } from 'react';
import { pretty } from '../node';
import type { Options } from '../shared/options';
import { readStream } from '../shared/read-stream.browser';
import { toPlainText } from '../shared/utils/to-plain-text';
import { importReactDOM } from './import-react-dom';

export const render = async (
  element: React.ReactElement,
  options?: Options,
) => {
  const suspendedElement = <Suspense>{element}</Suspense>;
  const reactDOMServer = await importReactDOM().then(
    // This is beacuse react-dom/server is CJS
    (m) => m.default,
  );
  console.log(reactDOMServer);

  const html = await new Promise<string>((resolve, reject) => {
    reactDOMServer
      .renderToReadableStream(suspendedElement, {
        onError(error: unknown) {
          reject(error);
        },
        progressiveChunkSize: Number.POSITIVE_INFINITY,
      })
      .then(readStream)
      .then(resolve)
      .catch(reject);
  });

  if (options?.plainText) {
    return toPlainText(html, options.htmlToTextOptions);
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, '')}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
};
