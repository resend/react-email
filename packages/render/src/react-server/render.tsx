import { convert } from 'html-to-text';
import { Suspense } from 'react';
import { experimental_renderToHTML as renderToHTML } from 'react-markup';
import { pretty } from '../node';
import type { Options } from '../shared/options';
import { plainTextSelectors } from '../shared/plain-text-selectors';

export const render = async (node: React.ReactNode, options?: Options) => {
  const suspendedElement = <Suspense>{node}</Suspense>;

  const html = await renderToHTML(suspendedElement);

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, '')}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
};
