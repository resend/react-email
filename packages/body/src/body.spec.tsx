import { render } from '@react-email/render';
import { Body } from './index';

describe('<Body> component', () => {
  it('renders children correctly', async () => {
    const testMessage = 'Test message';
    const html = await render(<Body>{testMessage}</Body>);
    expect(html).toContain(testMessage);
  });

  it('passes style and other props correctly', async () => {
    const style = { backgroundColor: 'red' };
    const html = await render(
      <Body data-testid="body-test" style={style}>
        Test
      </Body>,
    );
    expect(html).toContain('style="background-color:red"');
    expect(html).toContain('data-testid="body-test"');
  });

  it('renders correctly', async () => {
    const actualOutput = await render(<Body>Lorem ipsum</Body>);
    expect(actualOutput).toMatchSnapshot();
  });
});
