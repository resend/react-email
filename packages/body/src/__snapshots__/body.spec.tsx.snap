// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Body> component > renders correctly 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><body><table style="width:100%;height:100%" border="0" cellPadding="0" cellSpacing="0" role="presentation" align="center"><tbody><tr><td>Lorem ipsum</td></tr></tbody></table><!--/$--></body>"`;
